File: basic.js
class Point {
  constructor(x, y) {
    this.x = x;
    this.y = y;
  }
}

function distance(p1, p2) {
  const dx = p1.x - p2.x;
  const dy = p1.y - p2.y;
  return Math.sqrt(dx * dx + dy * dy);
}

function bruteForceClosestPair(points) {
  if (points.length < 2) return { distance: Infinity, pair: [] };
  
  let minDist = Infinity;
  let pair = [];
  
  for (let i = 0; i < points.length - 1; i++) {
    for (let j = i + 1; j < points.length; j++) {
      const d = distance(points[i], points[j]);
      if (d < minDist) {
        minDist = d;
        pair = [points[i], points[j]];
      }
    }
  }
  
  // Sort pair by x-coordinate
  pair.sort((a, b) => a.x - b.x);
  return { distance: minDist, pair };
}

function stripClosest(strip, d) {
  let minDist = d;
  let bestPair = [];
  
  for (let i = 0; i < strip.length; i++) {
    for (let j = i + 1; j < strip.length && (strip[j].y - strip[i].y) < minDist; j++) {
      const dist = distance(strip[i], strip[j]);
      if (dist < minDist) {
        minDist = dist;
        bestPair = [strip[i], strip[j]];
      }
    }
  }
  
  if (bestPair.length === 2) {
    bestPair.sort((a, b) => a.x - b.x);
  }
  
  return { distance: minDist, pair: bestPair };
}

function closestPairUtil(Px, Py) {
  const n = Px.length;
  
  if (n <= 3) {
    return bruteForceClosestPair(Px);
  }
  
  const mid = Math.floor(n / 2);
  const midPoint = Px[mid];
  
  const leftPx = Px.slice(0, mid);
  const rightPx = Px.slice(mid);
  
  // Create y-sorted arrays for left and right
  const leftPy = [];
  const rightPy = [];
  for (const p of Py) {
    if (p.x <= midPoint.x) {
      leftPy.push(p);
    } else {
      rightPy.push(p);
    }
  }
  
  const dl = closestPairUtil(leftPx, leftPy);
  const dr = closestPairUtil(rightPx, rightPy);
  
  let dmin = dl.distance < dr.distance ? dl : dr;
  
  // Build strip
  const strip = [];
  for (const p of Py) {
    if (Math.abs(p.x - midPoint.x) < dmin.distance) {
      strip.push(p);
    }
  }
  
  const stripResult = stripClosest(strip, dmin.distance);
  
  if (stripResult.distance < dmin.distance) {
    return stripResult;
  }
  
  return dmin;
}

function getClosestPair(points) {
  if (points.length < 2) {
    return { distance: Infinity, pair: [] };
  }
  
  // Sort by x and by y
  const Px = [...points].sort((a, b) => a.x - b.x);
  const Py = [...points].sort((a, b) => a.y - b.y);
  
  const result = closestPairUtil(Px, Py);
  
  // Make sure pair is always sorted by x
  if (result.pair.length === 2) {
    result.pair.sort((a, b) => a.x - b.x);
  }
  
  return result;
}

File: basic.py
from dataclasses import dataclass
from typing import List, Tuple
import math

@dataclass
class Point:
    x: float
    y: float

def distance(p1: Point, p2: Point) -> float:
    return math.sqrt((p1.x - p2.x)**2 + (p1.y - p2.y)**2)

def brute_force_closest_pair(points: List[Point]) -> Tuple[float, List[Point]]:
    if len(points) < 2:
        return float('inf'), []
    
    min_dist = float('inf')
    pair = []
    
    for i in range(len(points)):
        for j in range(i + 1, len(points)):
            d = distance(points[i], points[j])
            if d < min_dist:
                min_dist = d
                pair = [points[i], points[j]]
    
    pair.sort(key=lambda p: p.x)
    return min_dist, pair

def strip_closest(strip: List[Point], d: float) -> Tuple[float, List[Point]]:
    min_dist = d
    best_pair = []
    
    for i in range(len(strip)):
        for j in range(i + 1, len(strip)):
            if strip[j].y - strip[i].y >= min_dist:
                break
            dist = distance(strip[i], strip[j])
            if dist < min_dist:
                min_dist = dist
                best_pair = [strip[i], strip[j]]
    
    if len(best_pair) == 2:
        best_pair.sort(key=lambda p: p.x)
    
    return min_dist, best_pair

def closest_pair_util(Px: List[Point], Py: List[Point]) -> Tuple[float, List[Point]]:
    n = len(Px)
    
    if n <= 3:
        return brute_force_closest_pair(Px)
    
    mid = n // 2
    mid_point = Px[mid]
    
    left_Px = Px[:mid]
    right_Px = Px[mid:]
    
    left_Py = [p for p in Py if p.x <= mid_point.x]
    right_Py = [p for p in Py if p.x > mid_point.x]
    
    dl = closest_pair_util(left_Px, left_Py)
    dr = closest_pair_util(right_Px, right_Py)
    
    dmin, pair_min = (dl[0], dl[1]) if dl[0] < dr[0] else (dr[0], dr[1])
    
    strip = [p for p in Py if abs(p.x - mid_point.x) < dmin]
    
    ds, pair_s = strip_closest(strip, dmin)
    
    if ds < dmin:
        return ds, pair_s
    
    return dmin, pair_min

def get_closest_pair(points: List[Point]) -> dict:
    if len(points) < 2:
        return {"distance": float('inf'), "pair": []}
    
    Px = sorted(points, key=lambda p: p.x)
    Py = sorted(points, key=lambda p: p.y)
    
    dist, pair = closest_pair_util(Px, Py)
    
    if len(pair) == 2:
        pair.sort(key=lambda p: p.x)
    
    return {"distance": dist, "pair": pair}
